### PR TITLE
#61 Add subscription information to curve

### LINKED ISSUE #61: Include subscription informating in parse_curve

### PR DESCRIPTION
**Pull request by komape**
Closes #61

### CODE REVIEW

File: energyquantified/metadata/subscription.py [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+    def by_tag(tag):
+        """
+        Look up access by tag.
+
+        :param tag: An access tag
+        :type tag: str
+        :return: The access for the given tag
+        :rtype: SubscriptionAccess
+        """
+        return _access_lookup[tag.upper()]
```
Comment (fredriksvendsen) [inline energyquantified/metadata/subscription.py:101]: Get by tag in lower.

```python
return _access_lookup[tag.lower()]
```

File: energyquantified/metadata/subscription.py [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+            self.collection_perms = collection_perms
+        else:
+            self.collection = None
+            self.collection_perms = None
+
+    def __str__(self):
+        return self.label
+
+    def __repr__(self):
+        return f"<Subscription: \"{self.label}\", access={self.access}, type={self.type}>"
```
Comment (fredriksvendsen) [inline energyquantified/metadata/subscription.py:41]: Move the string representation into `__str__()` and change `__repr__()` to simply returning `self.__str__()`.

File: energyquantified/metadata/subscription.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+    def __init__(self, tag):
+        self.tag = tag
+        _type_lookup[tag.lower()] = self
+
+    def __str__(self):
+        return self.tag
+
+    def __repr__(self):
+        return f"<Type: \"{self.tag}\">"
```
Comment (fredriksvendsen) [inline energyquantified/metadata/subscription.py:133]: - [ ] Move string representation into `__str__()` and make `__repr__()` return `self.__str__()`
- [ ] Change from "<Type: .." to "<SubscriptionType: .."

```python
def __str__(self):
    return f"<SubscriptionType: {self.tag}>"

def __repr__(self):
    return self.__str__()
```

File: energyquantified/metadata/subscription.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+
+    def __init__(self, tag):
+        self.tag = tag
+        _collection_perm_lookup[tag.lower()] = self
+
+    def __str__(self):
+        return self.tag
+
+    def __repr__(self):
+        return f"<CollectionPerm: \"{self.tag}\">"
```
Comment (fredriksvendsen) [inline energyquantified/metadata/subscription.py:181]: - [ ] Move string representation into `__str__()` and make `__repr__()` return `self.__str__()`
- [ ] Change from "<CollectonPerm: .." to "<SubscriptionCollectionPerm: .."

```python
def __str__(self):
    return f"<SubscriptionCollectionPerm: {self.tag}>"

def __repr__(self):
    return self.__str__()
```

File: energyquantified/metadata/curve.py [hunk 5]
```diff
@@ -57,6 +57,8 @@ def __init__(self, name, curve_type=None, instance_issued_timezone=None,
         self.source = source
         #: The curve commodity (Power, Gas, etc.)
         self.commodity = commodity
+        #: The subscription for this curve, :py:class:`Subscription`.
```
Comment (fredriksvendsen) [inline energyquantified/metadata/curve.py:60]: #: The subscription for this curve, :py:class:`Subscription

Add "see":

#: The subscription for this curve, see :py:class:`Subscription

File: energyquantified/parser/metadata.py [hunk 6]
```diff
@@ -185,3 +191,28 @@ def parse_contract(json):
     raise ParseException(
         f"Unknown contract.type in JSON: {contract_type}"
     )
+
+
+def parse_subscription(json):
+    """
+    Parse a JSON response from the server into a Subscription object.
+    """
+
```
Comment (fredriksvendsen) [inline energyquantified/parser/metadata.py:200]: Remove blank line at 200.

File: energyquantified/parser/metadata.py [hunk 7]
```diff
@@ -6,6 +6,7 @@
     Curve, Instance, Area, DataType, CurveType, Place, PlaceType,
     ContractPeriod, ContinuousContract, SpecificContract, OHLCField
 )
+from ..metadata.subscription import SubscriptionAccess, SubscriptionType, Subscription, SubscriptionCollectionPerm
```
Comment (fredriksvendsen) [inline energyquantified/parser/metadata.py:9]: Import from metadata (after subscription models exported in `energyquantified.metadata.__init__.py`, described in review comment)

```python
from ..metadata import (
    Subscription,
    SubscriptionAccess,
    SubscriptionType,
    SubscriptionCollectionPerm,
)
```

**Review by fredriksvendsen — CHANGES_REQUESTED** [review]
- [x] Export new subscription models in energyquantified.metadata (see `exporting classes` described below)
- [ ] Describe `Subscription` in `docs/reference/packages.rst` in the `energyquantified.metadata` section.  Briefly mention the different models `SubscriptionAccess`, `SubscriptionCollectionPerm`, and `SubscriptionType`. If subscription is **only** related to `Curve` I think it makes sense to describe it under the `Curve` bulletpoint, otherwise add a new bullet point.
- [ ] Add `Subscription`, `SubscriptionAccess`, `SubscriptionCollectionPerm` and `SubscriptionType` in `docs/reference/reference.rst`
- [ ] Update CHANGELOG.rst
- [ ] Add subscription in string representation of `Curve` (see `Curve.__str__()` below)
- [ ] Breifly mention that curve's include subscription information under the `Curve search` section in `docs.userguide.metadata.rst`
- [ ] Describe the subscription object under the `Curve` section in `docs.userguide.terminology.rst`

## Exporting classes

Import in `energyquantified.metadata.__init__.py`:

```python
from .subscription import (
    Subscription,
    SubscriptionAccess,
    SubscriptionCollectionPerm,
    SubscriptionType,
)
```
This makes it so users can import `Subscription` like this:

`from energyquantified.metadata import Subscription`


And add in `_all_` in the same file:

```python
__all__ = [
    # ...
    # Subscription
    "Subscription",
    "SubscriptionAccess",
    "SubscriptionCollectionPerm",
    "SubscriptionType",
]
```

The `_all_` is used for when users want to import everything:

`from energyquantified.metadata import *`

## Curve.__str__()

NOTE: Await response on the questions below before doing changes.

I just noticed that `Curve.__str__()` only returns `self.name`, unlike `Curve.__repr__()`.
Add subscription in `__repr__`:

```python
def __repr__(self):
    return (
        f"<Curve: \"{self.name}\, "
        f"curve_type={self.curve_type}, "
        f"subscription={self.subscription}>"
    )
```

Question: Use `subscription={self.subscription}` or `subscription={self.subscription.access}`?

Question: We usually define string representations in `__str__()`, and define `__repr__()` to return `self.__str__()`. Do we want to change to this for `Curve`, or keep as is?

 @jonmd  please see the two questions above.

File: energyquantified/metadata/subscription.py [hunk 8]
```diff
@@ -0,0 +1,205 @@
...
+    #: Access is provided through partner agreements
+    INTERNAL = "INTERNAL"
+    #: Access is available at no cost
+    FREE = "FREE"
+    #: Missing access information
+    NONE = "NONE"
+
+    def __init__(self, tag):
+        self.tag = tag
+        _access_lookup[tag.upper()] = self
```
[the remark this answers is not in the document]
Comment (komape) [reply energyquantified/metadata/subscription.py:71]: But all the tags are uppercase. Don't we have to use `upper()` then?

File: energyquantified/metadata/subscription.py [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+    def is_valid_tag(tag):
+        """
+        Check whether an access tag exists or not.
+
+        :param tag: An access tag
+        :type tag: str
+        :return: True if it exists, otherwise False
+        :rtype: bool
+        """
+        return tag.lower() in _access_lookup[tag.upper()]
```
[the remark this answers is not in the document]
Comment (komape) [reply energyquantified/metadata/subscription.py:89]: See https://github.com/energyquantified/eq-python-client/pull/92#discussion_r1354861331

File: energyquantified/metadata/subscription.py (hunk 1 above)
Comment (komape) [reply energyquantified/metadata/subscription.py:101]: See https://github.com/energyquantified/eq-python-client/pull/92#discussion_r1354861331

File: energyquantified/metadata/subscription.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jonmd) [reply energyquantified/metadata/subscription.py:96]: We always convert to lowercase for lookups elsewhere in the code. So we should also do it here for consistency.

Also, @fredriksvendsen's comment above is relevant. Remove the index thing after `_access_lookup`:

```python
# Wrong
return tag.lower() in _access_lookup[tag.upper()]
# Correct
return tag.lower() in _access_lookup
```

File: energyquantified/metadata/subscription.py (hunk 1 above)
Comment (jonmd) [reply energyquantified/metadata/subscription.py:108]: See https://github.com/energyquantified/eq-python-client/pull/92#discussion_r1361660547

File: energyquantified/parser/metadata.py (hunk 7 above)
Comment (fredriksvendsen) [reply energyquantified/parser/metadata.py]: Not fixed

File: energyquantified/metadata/curve.py [hunk 10]
```diff
@@ -69,7 +71,8 @@ def __str__(self):
         return self.name
 
     def __repr__(self):
-        return f"<Curve: \"{self.name}\", curve_type={self.curve_type}>"
+        return f"<Curve: \"{self.name}\", curve_type={self.curve_type}, " \
+               f"subscription={self.subscription}>"
```
Comment (jonmd) [inline energyquantified/metadata/curve.py:75]: The current code prints something like:

```
<Curve: "DE Biomass Power Production MWh/h 15min Actual", curve_type=TIMESERIES, subscription=Power – Germany>
```

We want it to be. Whether you have `FREEMIUM`, `TRIAL`, `PAYING` or `BLOCKED` is the most important information in the `repr` on the `Curve` object:

```
<Curve: "DE Biomass Power Production MWh/h 15min Actual", curve_type=TIMESERIES, subscription=TRIAL>
```

File: docs/reference/packages.rst [hunk 11]
```diff
@@ -98,6 +98,16 @@ are quite many classes in ``energyquantified.metadata``:
...
+ * :py:class:`Subscription <energyquantified.metadata.Subscription>`: A class
+    describing your subscription for a given curve.
+ * :py:class:`SubscriptionAccess <energyquantified.metadata.SubscriptionAccess>`:
+   Enumerator of access levels for subscriptions.
+ * :py:class:`SubscriptionType <energyquantified.metadata.SubscriptionType>`:
+   Enumerator of subscription types.
+ * :py:class:`SubscriptionCollectionPerms
+    <energyquantified.metadata.SubscriptionCollectionPerms>`: Enumerator of
+    user's permission for a collection.
+
```
Comment (fredriksvendsen) [inline docs/reference/packages.rst:110]: Add a line shift between bullet points

File: docs/userguide/terminology.rst [hunk 12]
```diff
@@ -33,6 +36,46 @@ The ``curve_type`` tells you which operations you shall use to load data
...
+- ``package_area`` – A subscription tied to a combination of a specific package and area
+- ``private`` – A private subscription with restricted access
+
+**Subscription collection permissions**
+
+The user's permissions for a collection.
+
+- ``r`` – Read-only access
+- ``rw`` – Read-write access
+
```
Comment (fredriksvendsen) [inline docs/userguide/terminology.rst:78]: Add a blank line (two blank lines between sections)

**Review by jonmd — APPROVED** [review]

**Pull request merged**